File: src/main/kotlin/com/gh/wedding/domain/Notice.kt
package com.gh.wedding.domain

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Index
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.PrePersist
import jakarta.persistence.PreUpdate
import jakarta.persistence.Table
import java.time.LocalDateTime

@Entity
@Table(
    name = "notice",
    indexes = [
        Index(name = "idx_notice_status_start_end", columnList = "status,start_at,end_at"),
        Index(name = "idx_notice_banner", columnList = "is_banner,start_at"),
        Index(name = "idx_notice_created_at", columnList = "created_at"),
    ],
)
class Notice(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,

    @Column(nullable = false, length = 255)
    var title: String = "",

    @Column(nullable = false, columnDefinition = "TEXT")
    var content: String = "",

    @Column(name = "start_at", nullable = false)
    var startAt: LocalDateTime = LocalDateTime.now(),

    @Column(name = "end_at")
    var endAt: LocalDateTime? = null,

    @Column(name = "is_banner", nullable = false)
    var isBanner: Boolean = false,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    var status: NoticeStatus = NoticeStatus.DRAFT,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by", nullable = false)
    var createdBy: UserAccount? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "updated_by")
    var updatedBy: UserAccount? = null,

    @Column(name = "created_at")
    var createdAt: LocalDateTime? = null,

    @Column(name = "updated_at")
    var updatedAt: LocalDateTime? = null,
) {
    @PrePersist
    fun onCreate() {
        val now = LocalDateTime.now()
        createdAt = now
        updatedAt = now
    }

    @PreUpdate
    fun onUpdate() {
        updatedAt = LocalDateTime.now()
    }
}

File: frontend/src/types/notice.ts
export type NoticeStatus = "DRAFT" | "PUBLISHED" | "HIDDEN";

export type PagedResponse<T> = {
  content: T[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
  last: boolean;
};

export type NoticeSummary = {
  id: number;
  title: string;
  startAt: string;
  endAt?: string | null;
  isBanner: boolean;
};

export type NoticeDetail = {
  id: number;
  title: string;
  content: string;
  startAt: string;
  endAt?: string | null;
  isBanner: boolean;
  status: NoticeStatus;
  createdByUserId?: string | null;
  updatedByUserId?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
};

export type AdminNoticeSummary = {
  id: number;
  title: string;
  status: NoticeStatus;
  isBanner: boolean;
  startAt: string;
  endAt?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
};

export type NoticeUpsertRequest = {
  title: string;
  content: string;
  startAt: string;
  endAt?: string | null;
  isBanner: boolean;
  status: NoticeStatus;
};
